### PR TITLE
execution-core: added serde support for MoonlightTransactionEvent

### PR DESCRIPTION
**Pull request by d-sonuga**
This is to add the feature described in #2773.
I've only added serde support for a single event struct - `MoonlightTransactionEvent`.
I need some thoughts on whether what I have now is okay.

### CODE REVIEW

File: execution-core/Cargo.toml [hunk 1]
```diff
@@ -57,3 +62,5 @@ groth16 = [
 
 # Enables std feature for dusk-plonk
 std = ["dusk-plonk/std"]
+
+serde_support = ["serde", "bs58", "serde_json"]
```
Comment (Neotamandua) [inline execution-core/Cargo.toml:66]:
```suggestion
serde = ["serde", "bs58", "serde_json"]
```
Many crates expose serde support on their own crates through a feature flag named the same as the crate, so I would prefer to rename this to just "serde" as a flag

File: execution-core/src/transfer.rs [hunk 2]
```diff
@@ -505,20 +508,224 @@ pub struct PhoenixTransactionEvent {
...
+    )]
     pub receiver: Option<AccountPublicKey>,
     /// Transfer amount
     pub value: u64,
     /// The memo included in the transaction.
+    #[cfg_attr(
+        feature = "serde_support",
+        serde(with = "serde_support::bs58_serde")
+    )]
     pub memo: Vec<u8>,
```
Comment (Neotamandua) [inline execution-core/src/transfer.rs:531]: Memos should not be bs58 encoded. Prefer using hex for bytes and bs58 encoding only for the bytes of address types like AccountPublicKey

File: execution-core/src/transfer.rs [hunk 3]
```diff
@@ -505,20 +508,224 @@ pub struct PhoenixTransactionEvent {
...
     /// from the sender.
+    #[cfg_attr(
+        feature = "serde_support",
+        serde(with = "serde_support::pubk_u64_tuple")
+    )]
     pub refund_info: Option<(AccountPublicKey, u64)>,
 }
+
+#[cfg(feature = "serde_support")]
+mod serde_support {
```
Comment (Neotamandua) [inline execution-core/src/transfer.rs:544]: It probably makes sense to have this as a separate file. Then you can also import the modules in it that are likely to be reused (e.g., bs58_serde) on other structs located in other files.

File: core/src/serde_support.rs [hunk 4]
```diff
@@ -0,0 +1,1458 @@
...
+                Unexpected::Str(&s),
+                &"a non-empty string",
+            )
+        })?;
+        if last_char != 'n' {
+            return Err(SerdeError::invalid_value(
+                Unexpected::Str(&s),
+                &"a bigint ending with character 'n'",
+            ));
+        }
```
Comment (Neotamandua) [inline core/src/serde_support.rs:66]: I am not quite sure if we want to enforce this on the Deserialization part. Theoretically, we could support both without a problem. I will leave this comment here for further discussion.

File: core/tests/serde.rs [hunk 5]
```diff
@@ -0,0 +1,305 @@
...
+        receiver,
+        value: rng.next_u64(),
+    };
+
+    let ser = serde_json::to_string(&event).unwrap();
+    let deser = serde_json::from_str(&ser).unwrap();
+    assert_eq!(event, deser);
+}
+
+fn rand_note(rng: &mut StdRng) -> Note {
```
[the remark this answers is not in the document]
Comment (d-sonuga) [reply core/tests/serde.rs:222]: Yes, but doing that will result in ownership errors.

[the remark this answers is not in the document]
Comment (Neotamandua) [reply core/tests/serde.rs]: > Yes, but doing that will result in ownership errors.

```rust
fn rand_note() -> Note {
    let mut rng = StdRng::seed_from_u64(42);
    let pk = PhoenixPublicKey::from(&PhoenixSecretKey::random(&mut rng));
    let blinder = JubJubScalar::random(&mut rng);
    let sender_blinder = [
        JubJubScalar::random(&mut rng),
        JubJubScalar::random(&mut rng),
    ];
    Note::obfuscated(&mut rng, &pk, &pk, 42, blinder, sender_blinder)
}
```
This should work

File: core/src/serde_support.rs [hunk 6]
```diff
@@ -0,0 +1,1458 @@
...
+    }
+}
+
+impl<'de> Deserialize<'de> for StakeFundOwner {
+    fn deserialize<D: Deserializer<'de>>(
+        deserializer: D,
+    ) -> Result<Self, D::Error> {
+        struct StakeFundOwnerVisitor;
+
+        static VARIANTS: [&'static str; 2] = ["Account", "ContractId"];
```
Comment (Neotamandua) [inline core/src/serde_support.rs:103]:
```suggestion
        const VARIANTS: [&str; 2] = ["Account", "ContractId"];
```

any reason to have this as static?

Comment (d-sonuga) [reply core/src/serde_support.rs:103]: Actually, no.

File: core/Cargo.toml [hunk 7]
```diff
@@ -63,3 +69,15 @@ groth16 = [
 
 # Enables std feature for dusk-plonk
 std = ["dusk-plonk/std"]
+
+serde = [
```
[the remark this answers is not in the document]
Comment (moCello) [reply core/Cargo.toml:73]: actually I don't think we do. We don't use the `Desearialize` or `Serialize` trait functionalities for the jubjub-scalar or points here.

File: core/src/serde_support.rs [hunk 8]
```diff
@@ -0,0 +1,1457 @@
...
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use alloc::format;
+use alloc::string::ToString;
+use alloc::vec::Vec;
+
+use base64 as _;
```
Comment (moCello) [inline core/src/serde_support.rs:11]: This shouldn't be in here

File: core/src/serde_support.rs [hunk 9]
```diff
@@ -0,0 +1,1457 @@
...
+use base64::engine::general_purpose::STANDARD as BASE64_STANDARD;
+use base64::Engine;
+use dusk_bytes::Serializable;
+use hex as _;
```
Comment (moCello) [inline core/src/serde_support.rs:15]: this too

File: core/src/serde_support.rs [hunk 10]
```diff
@@ -0,0 +1,1457 @@
...
+use serde::de::{
+    EnumAccess, Error as SerdeError, MapAccess, Unexpected, VariantAccess,
+    Visitor,
+};
+use serde::ser::SerializeStruct;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+use serde_json as _;
```
Comment (moCello) [inline core/src/serde_support.rs:22]: this neither

File: core/Cargo.toml [hunk 11]
```diff
@@ -28,6 +28,12 @@ ark-bn254 = { workspace = true, features = ["curve"], optional = true }
 ark-relations = { workspace = true, optional = true }
 ark-serialize = { workspace = true, optional = true }
 
+# serde support dependencies
+serde = { workspace = true, features = ["derive"], optional = true }
+serde_json = { workspace = true, optional = true }
```
Comment (moCello) [inline core/Cargo.toml:33]: `serde_json` is a dev-dep.
The deny(unused_crate_dependencies) lint should catch this, but only if you don't use tricks like `use ... as _;'

File: core/src/serde_support.rs (hunk 10 above)
Comment (d-sonuga) [reply core/src/serde_support.rs:22]: The issue is that unused crate dependencies aren't allowed, so removing this will cause a compiler error.

File: core/Cargo.toml (hunk 7 above)
[the remark this answers is not in the document]
Comment (Neotamandua) [reply core/Cargo.toml:73]: You're right. This is only used indirectly by phoenix-core.

File: core/src/serde_support.rs [hunk 12]
```diff
@@ -0,0 +1,1454 @@
...
+    WithdrawEvent,
+};
+use crate::{
+    signatures::bls::PublicKey as AccountPublicKey, stake::StakeFundOwner,
+};
+use crate::{BlsScalar, String};
+
+// To serialize and deserialize u64s as big ints:
+// https://github.com/dusk-network/rusk/issues/2773#issuecomment-2519791322.
+struct Bigint(u64);
```
Comment (moCello) [inline core/src/serde_support.rs:35]: Please also add some tests for the `Bigint` serialization as well, especially edge-cases like `u64::MAX`

**Review by moCello — COMMENTED** [review]
Is there a reason why the traits are re-implemented for each type?
You might also be able to simply derive types that are composed of other types that implement `Serialize` and `Deserialize` already:
For example for `StakeFundOwner` adding this on top of its declaration will be enough:
```rust
#[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
pub enum StakeFundOwner {
...
}
```
This is probably true for many other types as well.

**Comment by d-sonuga** [comment]
> Is there a reason why the traits are re-implemented for each type? You might also be able to simply derive types that are composed of other types that implement `Serialize` and `Deserialize` already: For example for `StakeFundOwner` adding this on top of its declaration will be enough:
> 
> ```rust
> #[cfg_attr(feature = "serde", derive(serde::Serialize, serde::Deserialize))]
> pub enum StakeFundOwner {
> ...
> }
> ```
> 
> This is probably true for many other types as well.

The reason is to keep all implementations in one place, for ease of auditing. I think @Neotamandua mentioned something about this at some point before?

File: core/src/serde_support.rs (hunk 12 above)
Comment (d-sonuga) [reply core/src/serde_support.rs:35]: I'll wait for https://github.com/dusk-network/rusk/pull/3046#discussion_r1895820712 to be resolved first

**Review by moCello — APPROVED** [review]
Looks good from my side.
But please wait for a go from @ZER0 and @Neotamandua as well as they are the ones that can judge the serialization approach.

**Review by Neotamandua — DISMISSED** [review]

**Review by Neotamandua — APPROVED** [review]
LGTM

**Review by Neotamandua — APPROVED** [review]
